lib/redirect-pool.js

```diff
@@ -0,0 +1,149 @@
+'use strict'
+
+const { pipeline, request, stream } = require('./agent')
+const { InvalidArgumentError } = require('./core/errors')
+const util = require('./core/util')
+const Pool = require('./pool')
+
+const kOrigin = Symbol('origin')
+const kFollowedRedirections = Symbol('followed redirections')
+const kLeftRedirections = Symbol('left redirections')
+
+const redirectCodes = [300, 301, 302, 303, 307, 308]
+const defaultMaxRedirections = 10
+
+function parseMaxRedirections ({ maxRedirections = defaultMaxRedirections }) {
+  if (maxRedirections != null && (maxRedirections <= 0 || !Number.isInteger(maxRedirections))) {
+    throw new InvalidArgumentError('maxRedirections must be a positive number')
+  }
+
+  return maxRedirections
+}
+
+function redirectLocation (statusCode, headers, opts) {
+  if (opts[kLeftRedirections] < 0 || redirectCodes.indexOf(statusCode) === -1) {
+    return null
+  }
+
+  for (let i = 0; i < headers.length; i += 2) {
+    // Find the matching headers, then return its value
+    if (headers[i].length && headers[i].toLowerCase() === 'location') {
+      return headers[i + 1]
+    }
+  }
+
+  return null
+}
+
+// https://tools.ietf.org/html/rfc7231#section-6.4.4
+function shouldRemoveHeader (header, removeContent) {
+  const lcHeader = header.toLowerCase()
+
+  return lcHeader === 'host' || (removeContent && lcHeader.indexOf('content-') === 0)
+}
+
+// https://tools.ietf.org/html/rfc7231#section-6.4
+function cleanRequestHeaders (headers, removeContent) {
+  if (Array.isArray(headers)) {
+    for (let i = headers.length - 2; i >= 0; i -= 2) {
+      if (shouldRemoveHeader(headers[i], removeContent)) {
+        headers.splice(i, 2)
+      }
+    }
+  } else {
+    // IncomingHttpHeaders
+    for (const header of Object.keys(headers)) {
+      if (shouldRemoveHeader(header, removeContent)) {
+        headers[header] = undefined
+      }
+    }
+  }
+}
+
+class RedirectPool extends Pool {
+  constructor (origin, options) {
+    super(origin, options)
+
+    this[kOrigin] = origin
+  }
+
+  dispatch (opts, handler) {
+    const pool = this
+
+    // Compute the number of left redirects, if needed
+    if (!(kLeftRedirections in opts)) {
+      opts[kLeftRedirections] = parseMaxRedirections(opts)
```

Comment:
Don’t modify user passed/owned objects

Reply:
I'm not sure I understand your point here. When the pool is invoked via top-level agent, the opts are already modified (since we use destructuring and then rest assignment).

But if you're worried about direct use of the class ok, I'll change it.

Reply:
Done.